kite_test.go

```diff
@@ -32,6 +36,75 @@ func panicRegisterHandler(*protocol.RegisterResult) {
 	panic("this panic should be ignored")
 }
 
+func transportFromEnv() config.Transport {
+	env := os.Getenv("KITE_TRANSPORT")
+	tr, ok := config.Transports[env]
+	if env != "" && !ok {
+		panic(fmt.Errorf("transport %q doesn't exists", env))
+	}
+	return tr
+}
+
+func TestContext(t *testing.T) {
+	flag.Parse()
+
+	ch := make(chan int, 4) // checkpoints, to ensure correct control flow
+
+	k := New("server", "0.0.1")
+	k.Config.DisableAuthentication = true
+	k.Config.Port = 3333
+	k.Config.Transport = transportFromEnv()
+	k.HandleFunc("longrunning", func(r *Request) (interface{}, error) {
+		ch <- 2
+
+		go func() {
+			<-r.Context.Done()
+			ch <- 4
+		}()
+		return nil, nil
+	})
+	go k.Run()
+	<-k.ServerReadyNotify()
+	defer k.Close()
+
+	c := New("client", "0.0.1").NewClient("http://127.0.0.1:3333/kite")
+	if err := c.Dial(); err != nil {
+		t.Fatalf("Dial()=%s", err)
+	}
+
+	ch <- 1
+
+	if _, err := c.TellWithTimeout("longrunning", *timeout); err != nil {
+		t.Fatalf("TellWithTimeout()=%s", err)
+	}
+
+	ch <- 3
+
+	c.Close()
+
+	var got []int
+	want := []int{1, 2, 3, 4}
+	timeout := time.After(2 * time.Second)
+
+collect:
```

Comment:
could we not use labels if not critical?

Reply:
```go
for len(got) != len(want) {
 	select {
 	case i := <-ch:
 		got = append(got, i)
 	case <-timeout:
 		t.Fatal("timed out collecting checkpoints")
 	}
 }
```

Reply:
Done.